src/promise-utils.ts

```diff
@@ -0,0 +1,41 @@
+export function timeoutPromise<T>(ms: number, promise: Promise<T>, message = `timed out after ${ms}ms`): Promise<T> {
+    return new Promise(function (resolve, reject) {
+        setTimeout(() => reject(new Error(message)), ms);
+        promise.then(resolve).catch(reject);
+    });
+}
+
+export function delayedPromise(delay: number): Promise<void> {
+    return new Promise(resolve => setTimeout(resolve, delay));
+}
+
+export interface RetryPromiseOptions {
+    interval: number;
+    retries: number;
+    timeout?: number;
+}
+
+export async function retryPromise<T>(promiseProvider: () => Promise<T>, {interval, retries, timeout}: RetryPromiseOptions): Promise<T> {
+    let aborted = false;
+    let lastError: Error;
+    return new Promise<T>(async (resolve, reject) => {
```

Comment:
This is called the explicit construction antipattern

Reply:
@benjamingr 
Indeed, this *felt* ugly while writing. I already have a refactor going on, where this ugliness is gone and proper unit tests are in place, verifying the mini behaviors. This was most relevant for the retry helper.